internal/directives/engine.go

```diff
@@ -6,42 +6,30 @@ import (
+	// Promote executes the provided list of PromotionSteps in sequence and
+	// returns a PromotionResult that aggregates the results of all steps.
+	Promote(context.Context, PromotionContext, []PromotionStep) (PromotionResult, error)
+	// CheckHealth executes the provided list of HealthCheckSteps in sequence and
+	// and returns a HealthCheckResult that aggregates the results of all steps.
+	CheckHealth(context.Context, HealthCheckContext, []HealthCheckStep) kargoapi.Health
 }
 
-// PromotionContext is the context of the Promotion that is being executed by
-// the Engine.
-type PromotionContext struct {
-	// WorkDir is the working directory to use for the Promotion.
-	WorkDir string
-	// Project is the Project that the Promotion is associated with.
-	Project string
-	// Stage is the Stage that the Promotion is targeting.
-	Stage string
-	// FreightRequests is the list of Freight from various origins that is
-	// requested by the Stage targeted by the Promotion. This information is
-	// sometimes useful to Steps that reference a particular artifact and, in the
-	// absence of any explicit information about the origin of that artifact, may
-	// need to examine FreightRequests to determine whether there exists any
-	// ambiguity as to its origin, which a user may then need to resolve.
-	FreightRequests []kargoapi.FreightRequest
-	// Freight is the collection of all Freight referenced by the Promotion. This
-	// collection contains both the Freight that is actively being promoted as
-	// well as any Freight that has been inherited from the target Stage's current
-	// state.
-	Freight kargoapi.FreightCollection
-}
-
-// Step is a single step that should be executed by the Engine.
-type Step struct {
-	// Directive is the name of the directive to execute for this step.
-	Directive string
-	// Alias is an optional alias for the step, which can be used to
-	// refer to its results in subsequent steps.
-	Alias string
-	// Config is a map of configuration values that can be passed to the step.
-	Config Config
+// Directive is an interface for components that implement the logic for
+// execution of the individual PromotionSteps of a user-defined promotion
+// process. Implementations may optionally define corresponding health check
+// procedures as well.
+type Directive interface {
+	// Name returns the name of the Directive.
+	Name() string
+	// RunPromotionStep executes an individual step of a user-defined promotion
+	// process using the provided PromotionStepContext. Implementations may
+	// indirectly modify that context through the returned PromotionStepResult to
+	// allow subsequent promotion steps to access the results of its execution.
+	RunPromotionStep(context.Context, *PromotionStepContext) (PromotionStepResult, error)
+	// RunHealthCheckStep executes a health check using the provided
+	// HealthCheckStepContext.
+	RunHealthCheckStep(context.Context, *HealthCheckStepContext) HealthCheckStepResult
```

Comment:
Wonder if we should split this up into e.g. `Promotable` and `Healthable` interfaces, which would allow "partial" implementations.

Reply:
That's a great idea. Consider it done.